### DISCUSSION TITLE
Why not `const` for `Coefficient` member functions?

### DISCUSSION THREAD

**Discussion by j-signorelli** [mfem/mfem#4171] [Q&A] [answered]
Hello,

There have been many times while using MFEM that I have wanted to pass a `Coefficient`, `VectorCoefficient`, or `MatrixCoefficient` as `const` to a class such as an operator, where I want to be explicit in the code that the class cannot alter the coefficient object but may use it.

However, I have noticed that all of the member functions of those three base classes (minus some additional ones in `MatrixCoefficient`) are not declared as `const`, and was wondering if there was any particular reason for this? For example, I don't immediately see how calling `Eval` generally may change the state of these coefficient objects, yet it is still not `const`. My guess is that this is why anything that uses a coefficient such as `GridFunction::Project___` and LinearForm/BilinearFormIntegrator constructors must take a non-`const` coefficient reference.

I was just wondering if there was any particular reason for this that I am missing?

Thank you!

**Comment by v-dobrev** [comment, marked as answer]
One perspective: using a `Coefficient` may require calling its `SetTime()` method which does alter the `Coefficient` state, so time-dependent `Coefficient`s need to be passed as non-const.

That said, I think that this interface dates back to the first MFEM release when the const-ness of methods was not considered thoroughly.

It is probably possible to change these methods to be const -- I'm not sure if this is a simple change local to the coefficient classes or if it needs to propagate to other classes too. If you feel strongly about it and you want to explore the implication of a such change, feel free to try it, and propose a PR if the changes are not too pervasive. The main thing will be to ensure that the changes are backward-compatible.

**Reply by j-signorelli** [reply]
Got it, thank you! There definitely would be a good bit of propagation through (like with methods for projecting coefficients and coefficient arguments for integrators) but I am interested in giving it a go and seeing how pervasive it is. I can't immediately think of any issues with backward compatibility but will keep that in mind.

**Reply by paul-hilscher** [reply]
> I am interested in giving it a go and seeing how pervasive it is.

that would be amazing ! Currently I am using `const_cast` when calling `Eval` in my code and would love to see this gone. Let me know if I can help in any way.

**Reply by j-signorelli** [reply]
PR submitted here: #4196 @paul-hilscher @v-dobrev
